WIP: Some test improvements

jat255 opened the pull request:
This is a spot for some test improvements I'm working on.

jat255 commented:
I'm wondering, can these checks ever fail? I can't write a test that will trigger them, because `factors` is defined by `get_decomposition_factors()`, which will fail internally if the `learning_results.factors` contains anything other than a properly shaped numpy array.

[L116-L132](https://github.com/AndrewHerzing/hyperspy/blob/MultivariateCurveResolution/hyperspy/learn/mcr.py#L116-L132):

```python
    # Check factors
    if not isinstance(factors, BaseSignal):
        raise ValueError(
            f"`factors` must be a BaseSignal instance, but an object of type "
            f"{type(factors)} was provided.")

    # Check factor dimensions
    if factors.axes_manager.navigation_dimension != 1:
        raise ValueError(f"`factors` must have navigation dimension"
                         f"equal one, but the navigation dimension "
                         f"of the given factors is "
                         f"{factors.axes_manager.navigation_dimension}.")
    elif factors.axes_manager.navigation_size < 2:
        raise ValueError(f"`factors` must have navigation size"
                         f"greater than one, but the navigation "
                         f"size of the given factors is"
                         f"{factors.axes_manager.navigation_size}.")
```

jat255 commented:
I realize now that this was copied from the blind source separation code... I'm still wondering though.

jat255 commented:
Ok, I see what's going on... In the original `bss` code ([here](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/hyperspy/learn/mva.py#L552-L570)), you can supply `factors` as a parameter, to provide your own inputs. So those consistency checks are in place to prevent people from giving a bogus input. 

Currently, the MCR only operates on a previous decomposition. We should either:

1) remove these checks because they're always valid in the current code (not recommended), or
2) allow users to supply their own factors to the MCR algorithm, if desired, like in the BSS code (recommended)

What do you think?

jat255 commented:
Another issue I spot:

The `mcrals` function in `mcr.py` has `mask` and `compute` as parameters, but from within the `decomosition` function, these are always set to `None` and `False`. Was your intention to make these options available if someone wanted to use the `mcr.mcrals` function directly, or are these vestigial from the conversion from the BSS code?

AndrewHerzing commented:
I think allowing the user to supply their own factors is the way to go and a good suggestion.

As for the mask and compute functions, these should probably be activated since I think they are available in all other MVA options.

jat255 commented:
Ok, still working on this, but made some changes:

1) Parameterized the use of `normalize_poissonian_noise` so we get better coverage by calling each test both with and without normalization
2) removed component number check, since there's a hard check on the presence of the `output_dimension` parameter in `mva.py`
3) removed the check for NaNs in the loading maps, since I think those will always be well-formed even when there are zero values (in my testing, at least)
4) made some exception assertions more specific with the `match` parameter

jat255 commented:
> I think allowing the user to supply their own factors is the way to go and a good suggestion.
> 
> As for the mask and compute functions, these should probably be activated since I think they are available in all other MVA options.

Agreed on both fronts. I'll look into adding them in

jat255 commented:
> I think allowing the user to supply their own factors is the way to go and a good suggestion.

If we're allowing pre-defined factors, we should also allow pre-defined loadings, correct (since the user can do spatial or spectral simplicity)?
